File: java/starter/lp/src/main/java/network/t0/pay/lp/internal/FiatSettlementSent.java
package network.t0.pay.lp.internal;

import io.grpc.StatusRuntimeException;
import network.t0.pay.proto.tzero.v1.pay.Decimal;
import network.t0.pay.proto.tzero.v1.pay.FiatSettlementSentRequest;
import network.t0.pay.proto.tzero.v1.pay.FiatSettlementSentResponse;
import network.t0.pay.proto.tzero.v1.pay.LpServiceGrpc;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.Instant;
import java.util.List;
import java.util.concurrent.TimeUnit;

/**
 * §10 FiatSettlementSent — you released a bank transfer to the acquirer covering a
 * batch of your executions.
 *
 * <p>Self-initiated: nothing from t-0 tells you to pay. You decide when to batch and
 * send, then report it here. t-0 relays it to the acquirer as §11 so it knows which
 * reference to watch for, and the intent settles on the acquirer's §12.
 *
 * <p>Idempotency key: {@code bankTransferRef}, your reference on the transfer,
 * unique per LP. One ref is one real transfer — a second transfer is a new ref.
 *
 * <p>{@code settlementAmount} must equal the sum of the covered executions' local
 * amounts, and one transfer credits one acquirer.
 */
public final class FiatSettlementSent {

    private static final Logger log = LoggerFactory.getLogger(FiatSettlementSent.class);

    /**
     * Longer than the 10s default Main installs: the money already left your bank, so
     * it is worth waiting rather than turning a transfer that landed into an
     * {@link Outcome.Unknown} you have to reconcile.
     */
    private static final int TIMEOUT_SECONDS = 15;

    public static Outcome<FiatSettlementSentResponse.Accepted> report(
            LpServiceGrpc.LpServiceBlockingStub t0,
            String bankTransferRef,
            List<Long> settledExecutionIds,
            String localCurrency,
            Decimal settlementAmount,
            String destinationAccount,
            Instant settledAt) {

        FiatSettlementSentRequest request = FiatSettlementSentRequest.newBuilder()
                .setBankTransferRef(bankTransferRef)
                .addAllSettledExecutionIds(settledExecutionIds)
                .setLocalCurrency(localCurrency)
                .setSettlementAmount(settlementAmount)
                .setDestinationAccount(destinationAccount)
                .setSettledAt(Times.from(settledAt))
                .build();

        try {
            FiatSettlementSentResponse response =
                    t0.withDeadlineAfter(TIMEOUT_SECONDS, TimeUnit.SECONDS).fiatSettlementSent(request);

            switch (response.getResultCase()) {
                case ACCEPTED -> {
                    log.info("§10 accepted: ref={} {} {} covering {}",
                            bankTransferRef, Decimals.format(settlementAmount), localCurrency,
                            settledExecutionIds);
                    return new Outcome.Accepted<>(response.getAccepted());
                }
                case REJECTED -> {
                    // EXECUTION_UNKNOWN / EXECUTION_ALREADY_COVERED / CURRENCY_MISMATCH /
                    // AMOUNT_MISMATCH / DESTINATION_MISMATCH / ACQUIRER_MIXED.
                    // The money already moved: the fix is in what you reported, not another
                    // transfer. Correct the fields and resend the same ref.
                    FiatSettlementSentResponse.Rejected rejected = response.getRejected();
                    log.warn("§10 rejected for ref {}: {} (failing executions {})",
                            bankTransferRef, rejected.getReason(), rejected.getFailingExecutionIdsList());
                    return new Outcome.Rejected<>(
                            rejected.getReason().name(), rejected.getFailingExecutionIdsList());
                }
                default -> {
                    return new Outcome.Rejected<>("response carried no result variant");
                }
            }
        } catch (StatusRuntimeException e) {
            log.error("§10 failed for ref {}: {}", bankTransferRef, e.getStatus());
            return new Outcome.Unknown<>(e.getStatus().toString());
        }
    }

    private FiatSettlementSent() {
    }
}

File: node/starter/lp/src/internal/fiat_settlement_sent.ts
import { timestampFromDate } from "@bufbuild/protobuf/wkt";
import {
  type Client,
  type Decimal,
  type LpService,
  type FiatSettlementSentResponse_Accepted,
  FiatSettlementSentResponse_Rejected_Reason,
} from "@t-0/usdt-pay-sdk";
import { decimalToString } from "./decimals.js";
import { accepted, noResultVariant, outcomeFromError, rejected, type Outcome } from "./outcome.js";

/**
 * Longer than the 10s the other calls use: the transfer is already sent, so
 * it is worth waiting rather than turning a settlement that landed into an `unknown`
 * you have to reconcile.
 */
const TIMEOUT_MS = 15_000;

/**
 * FiatSettlementSent — you wired local fiat to the acquirer over bank rails;
 * t-0 verifies the covered executions.
 *
 * Idempotency key: `bankTransferRef`, your own id, unique per LP. It identifies one
 * real bank transfer: a second wire is a new settlement under a new ref, never a
 * correction of an old one.
 *
 * `FAILED_PRECONDITION` means an execution in the batch has no durable result yet —
 * retry the same request. The contract says the same `bankTransferRef` with the same
 * content is safe to resend.
 */
export async function reportFiatSettlementSent(
  t0: Client<typeof LpService>,
  settlement: {
    bankTransferRef: string;
    settledExecutionIds: bigint[];
    localCurrency: string;
    settlementAmount: Decimal;
    settledAt: Date;
  },
): Promise<Outcome<FiatSettlementSentResponse_Accepted>> {
  try {
    const response = await t0.fiatSettlementSent(
      {
        bankTransferRef: settlement.bankTransferRef,
        settledExecutionIds: settlement.settledExecutionIds,
        localCurrency: settlement.localCurrency,
        settlementAmount: settlement.settlementAmount,
        settledAt: timestampFromDate(settlement.settledAt),
      },
      { timeoutMs: TIMEOUT_MS },
    );

    switch (response.result.case) {
      case "accepted":
        console.log(
          `FiatSettlementSent accepted: ref=${settlement.bankTransferRef} ${decimalToString(settlement.settlementAmount)} ${settlement.localCurrency} covering ${settlement.settledExecutionIds}`,
        );
        return accepted(response.result.value);

      case "rejected": {
        const { reason } = response.result.value;
        const name = FiatSettlementSentResponse_Rejected_Reason[reason] ?? String(reason);
        console.warn(
          `FiatSettlementSent rejected for ref ${settlement.bankTransferRef}: ${name}`,
        );
        return rejected(name);
      }

      default:
        return noResultVariant();
    }
  } catch (error) {
    // Never wire a second transfer to "retry" — resend this same ref.
    console.error(`FiatSettlementSent failed for ref ${settlement.bankTransferRef}:`, error);
    return outcomeFromError(error);
  }
}
